### PR TITLE
various updates to the logging page

### PR DESCRIPTION
**Pull request by devoncarew**
Various updates to the logging page:
- make the table rows more compact
- change the table to render in monospaced font
- change the logging 'kind' column to render with a colored background
- allow for specific rows to have custom rendering
- use the custom rendering to display frame duration as different length bars for `flutter.frame` events
- have the logging details area display across the bottom of the page
- fix an issue with the logging details area not using the full width of its area
- add a new `'Search'` text field to the logging page, and use it to filter the displayed rows
- some changes to `LoggingController` to allow us to test just the controller (new tests in `logging_controller_test.dart`)
- added a bunch of todos for future work and improvements

Screen Shot 2020-03-15 at 6 05 59 AM

Screen Shot 2020-03-15 at 6 06 26 AM

Screen Shot 2020-03-15 at 6 06 41 AM

### CODE REVIEW

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 1]
```diff
@@ -329,17 +334,18 @@ class _TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
...
-      @required this.rowBuilder,
-      this.startAtBottom = false})
-      : super(key: key);
+  const _Table({
+    Key key,
+    @required this.itemCount,
+    @required this.columns,
+    @required this.columnWidths,
+    @required this.rowBuilder,
+    this.populateInReverse = false,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: should we just name this `reverse` to match the ListView `reverse` named paramater?

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 2]
```diff
@@ -613,8 +629,17 @@ class _TableRowState<T> extends State<TableRow<T>>
     }
   }
 
+  TextStyle _fixedFontStyle(BuildContext context) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: nit:  consider moving this helper to src/flutter/theme.dart

File: packages/devtools_app/lib/src/logging/flutter/logging_screen.dart [hunk 3]
```diff
@@ -15,6 +17,13 @@ import '../../table_data.dart';
 import '../../ui/flutter/service_extension_widgets.dart';
 import '../logging_controller.dart';
 
+// TODO(devoncarew): Show rows starting from the top (and have them grow down).
+// TODO(devoncarew): We should keep new items visible (if the last item was
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/flutter/logging_screen.dart]: I thought the existing implementation already keeps new items visible if the last item was already visible.

File: packages/devtools_app/lib/src/logging/flutter/logging_screen.dart [hunk 4]
```diff
@@ -37,38 +46,77 @@ class LoggingScreenBody extends StatefulWidget {
...
+        mainAxisAlignment: MainAxisAlignment.start,
         children: [
           RaisedButton(
             child: const Text('Clear logs'),
             onPressed: _clearLogs,
           ),
+          const Spacer(),
+          Container(
+            width: 200.0,
+            height: 36.0,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/flutter/logging_screen.dart]: do you need to specify height: 36? Can you achieve the same effect by specifying cross-axis-alignment.

**Comment by jacob314** [comment]
Looks great in the dark theme! What does it look like in the light theme? You may need to use ThemedColor to specify some dark theme color defaults.

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 1 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/flutter/table.dart]: Sure; done.

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 2 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/flutter/table.dart]: done

File: packages/devtools_app/lib/src/logging/flutter/logging_screen.dart (hunk 3 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/logging/flutter/logging_screen.dart]: It does, by reversing scrolling and table entry.

This isn't ideal though:
- the rows start out drawing on the bottom for empty tables - this looks odd, esp. when the table headers are ~800 px higher up
- on new items, the table always scrolls, even if you're looking at items higher up (and have manually adjusted the scroll position). It's really hard to click on items if you have an app which logs regularly

I think we want to re-work things here; remove the `reverse` code, auto-scroll if the scroll position is at the bottom, and un-dock from auto-scrolling if the user has manually scrolled the last back up a bit.

File: packages/devtools_app/lib/src/logging/flutter/logging_screen.dart (hunk 4 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/logging/flutter/logging_screen.dart]: I'll try this a few different ways.

Comment (devoncarew) [reply packages/devtools_app/lib/src/logging/flutter/logging_screen.dart]: The default height of the text field is too large - it looks out of place with the other elements in the toolbar.

**Comment by devoncarew** [comment]
Thanks for the review! I believe the comments are addressed; ptal.

**Pull request merged**